### DISCUSSION TITLE
Support WebAssembly

### DISCUSSION THREAD

**Discussion by pi0** [nuxt/nuxt#17124] [RFCs]
Supporting WebAssembly (WASM) in nitro and possibly also App (webpack/vite) can be a game-changer for Nuxt!

Currently, we have a wide ecosystem that have runtime support of wasm:
- Node.js: https://nodejs.org/api/wasi.html
- Cloudflare workers: https://blog.cloudflare.com/webassembly-on-cloudflare-workers/
- Deno: https://deno.land/manual@v1.13.2/getting_started/webassembly
- Browser: https://developer.mozilla.org/en-US/docs/WebAssembly
- Lambda edge: https://markentier.tech/posts/2021/01/rust-wasm-on-aws-lambda-edge/

In addition to libraries that are pre-built for wasm, it can bring the possibility to integrate with non-js languages such as:

- Python (https://github.com/wasmerio/wasmer-python and https://github.com/pyodide/pyodide)
- PHP (https://github.com/wasmerio/wasmer-php)
- Go Lang (https://github.com/wasmerio/wasmer-go)
- Rust [official!] (https://rustwasm.github.io/docs/book/)
- Dotnet C# (https://dotnet.microsoft.com/apps/aspnet/web-apps/blazor)

Some steps to take:

- [x] Basic support with [@rollup/plugin-wasm](https://www.npmjs.com/package/@rollup/plugin-wasm) for nitro to inline WASM binaries (#693)
- [ ]  Move wasm binaries to individual chunks for worker target (#485, #484)
- [ ] Emit `.wasm` chunks for environments that support it instead of base64 encoding
- [ ] Experiment language extension integrations (to build source to wasm based on language handler)

**Comment by Arideno** [comment]
Hello. Is there any tutorial how to connect wasm to nuxt 3 project?

**Reply by cpreston321** [reply]
@Arideno Here: https://github.com/nuxt/framework/tree/main/examples/experimental/wasm

Idk if this a tutorial but more an example to work from. Hopefully this helps!
